src/core/meshcop/joiner.cpp

```diff
@@ -129,7 +129,10 @@ otError Joiner::Start(const char *     aPskd,
     if (error != OT_ERROR_NONE)
     {
         otLogInfoMeshCoP("Error %s while starting joiner", otThreadErrorToString(error));
-        FreeJoinerFinalizeMessage();
+        if (mState == OT_JOINER_STATE_IDLE)
```

Comment:
How about moving this check into `FreeJoinerFinalizeMessage()`? This change would also require moving the call to `FreeJoinerFinalizeMessage()` in `Joiner::Finish()` to after the `SetState(OT_JOINER_STATE_IDLE)` call. Thoughts?

Reply:
Makes sense to me.
Done.